go/vt/vtgate/engine/dbddl.go

```diff
@@ -0,0 +1,202 @@
+	queryTimeout int
+
+	noInputs
+	noTxNeeded
+}
+
+// NewDBDDL creates the engine primitive
+// `create` will be true for CREATE, and false for DROP
+func NewDBDDL(dbName string, create bool, timeout int) *DBDDL {
+	return &DBDDL{
+		name:         dbName,
+		create:       create,
+		queryTimeout: timeout,
+	}
+}
+
+// RouteType implements the Primitive interface
+func (c *DBDDL) RouteType() string {
+	if c.create {
+		return "CreateDB"
+	}
+	return "DropDB"
+}
+
+// GetKeyspaceName implements the Primitive interface
+func (c *DBDDL) GetKeyspaceName() string {
+	return c.name
+}
+
+// GetTableName implements the Primitive interface
+func (c *DBDDL) GetTableName() string {
+	return ""
+}
+
+// Execute implements the Primitive interface
+func (c *DBDDL) Execute(vcursor VCursor, _ map[string]*querypb.BindVariable, _ bool) (*sqltypes.Result, error) {
+	name := vcursor.GetDBDDLPluginName()
+	plugin, ok := databaseCreatorPlugins[name]
+	if !ok {
+		log.Errorf("'%s' database ddl plugin is not registered. Falling back to default plugin", name)
+		plugin = databaseCreatorPlugins[defaultDBDDLPlugin]
+	}
+	if c.queryTimeout != 0 {
+		cancel := vcursor.SetContextTimeout(time.Duration(c.queryTimeout) * time.Millisecond)
+		defer cancel()
+	}
+
+	if c.create {
+		return c.createDatabase(vcursor, plugin)
+	}
+
+	return c.dropDatabase(vcursor, plugin)
+}
+
+func (c *DBDDL) createDatabase(vcursor VCursor, plugin DBDDLPlugin) (*sqltypes.Result, error) {
+	ctx := vcursor.Context()
+	err := plugin.CreateDatabase(ctx, c.name)
+	if err != nil {
+		return nil, err
+	}
+	var destinations []*srvtopo.ResolvedShard
+	for {
+		// loop until we have found a valid shard
+		destinations, _, err = vcursor.ResolveDestinations(c.name, nil, []key.Destination{key.DestinationAllShards{}})
+		if err == nil {
+			break
+		}
+		select {
+		case <-ctx.Done(): //context cancelled
+			return nil, vterrors.Errorf(vtrpc.Code_DEADLINE_EXCEEDED, "could not validate created database")
+		case <-time.After(500 * time.Millisecond): //timeout
+		}
+	}
+	var queries []*querypb.BoundQuery
+	for range destinations {
+		queries = append(queries, &querypb.BoundQuery{
+			Sql:           "select 42 from dual where null",
+			BindVariables: nil,
+		})
+	}
+
+	for {
+		_, errors := vcursor.ExecuteMultiShard(destinations, queries, false, true)
+
+		noErr := true
+		for _, err := range errors {
+			if err != nil {
+				noErr = false
+				select {
+				case <-ctx.Done(): //context cancelled
+					return nil, vterrors.Errorf(vtrpc.Code_DEADLINE_EXCEEDED, "could not validate created database")
+				case <-time.After(500 * time.Millisecond): //timeout
+				}
+				break
+			}
+		}
+		if noErr {
+			break
+		}
+	}
+	return &sqltypes.Result{RowsAffected: 1}, nil
+}
+
+func (c *DBDDL) dropDatabase(vcursor VCursor, plugin DBDDLPlugin) (*sqltypes.Result, error) {
+	ctx := vcursor.Context()
+	err := plugin.DropDatabase(ctx, c.name)
+	if err != nil {
+		return nil, err
+	}
+	for vcursor.FindKeyspace(c.name) {
```

Comment:
It looks like `vcursor.FindKeyspace()` does a direct topo call, which checks the source of truth, but it doesn't check whether this particular vtgate instance has observed the change yet.

My goal was to guarantee that an immediate `CREATE DATABASE x` after `DROP DATABASE x` on the same vtgate session will succeed. It looks like `CREATE DATABASE` checks `vschema.KeyspaceExists(ksName)` to determine whether the keyspace exists. Wouldn't we want to check the same thing to guarantee that vtgate will allow it?